airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/components/debezium/DebeziumComponent.kt

```diff
@@ -0,0 +1,389 @@
+     *
+     * The [debeziumProperties] property map is used to initialize the Debezium engine.
+     *
+     * The [lsnMapper] interface is used to compare relative CDC progress without knowledge of how
+     * the progress itself is encoded.
+     *
+     * The rest are termination criteria:
+     * - [upperBound] is the targeted upper bound of CDC progress, [collect] stops when it's reached
+     * or exceeded,
+     * - [maxRecords] is the threshold of collected [Record]s at or beyond which [collect] stops,
+     * - [maxRecordBytes] is the threshold of collected JSON [Record] bytes at or beyond which
+     * [collect] stops,
+     * - [maxTime] is the threshold of collected JSON-serialized [Record] bytes at or beyond which
+     * [collect] stops,
+     */
+    data class Config(
+        val debeziumProperties: Properties,
+        val lsnMapper: LsnMapper<*>,
+        val upperBound: State.Offset,
+        val maxRecords: Long,
+        val maxRecordBytes: Long,
+        val maxTime: Duration,
+    ) {
+
+        /** Name of the Debezium instance defined by this [Config]. */
+        val debeziumName: String
+            get() = debeziumProperties.getProperty("name")
+
+        /**
+         * [LsnMapper] abstracts the mapping from [State.Offset] and [Record] to a
+         * database-dependent LSN value. All that matters as far as the [DebeziumComponent] is
+         * concerned is that the values are [Comparable]. Its clients by design are the
+         * [checkWithinBound] and [checkProgress] methods.
+         */
+        interface LsnMapper<T : Comparable<T>> {
+
+            /** Maps a [State.Offset] to an LSN value. */
+            fun get(offset: State.Offset): T
+
+            /** Maps a [Record] to an LSN value. */
+            fun get(record: Record): T?
+
+            /** Maps a [SourceRecord] to an LSN value. */
+            fun get(sourceRecord: SourceRecord): T?
+        }
+
+        /**
+         * Returns true iff the Debezium [Record] is within bounds as defined by [upperBound]. The
+         * Debezium [SourceRecord] accompanying the [Record] is preferred, but this object is
+         * internal to the Debezium Engine and therefore may not always be available.
+         */
+        fun checkWithinBound(record: Record, sourceRecord: SourceRecord?): Boolean {
+            val lsn = (sourceRecord?.let(typedLsnMapper()::get) ?: typedLsnMapper().get(record))
+            return if (lsn == null) true else lsn < upperBoundLsn
+        }
+
+        /** Returns true iff progress has been made between [before] and [after]. */
+        fun checkProgress(before: State, after: State): Boolean =
+            typedLsnMapper().get(before.offset) < typedLsnMapper().get(after.offset)
+
+        /**
+         * Convenience function that allows us to ignore the type parameter. The compiler can't tell
+         * that LsnMapper<*> methods all return the same (albeit unknown) type.
+         */
+        @Suppress("UNCHECKED_CAST")
```

Comment:
isn't this a sign that the outer type needs to be parameterized?

Reply:
I deliberately don't want to leak the LSN type.